lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs

```diff
@@ -371,19 +371,36 @@ role = fromMaybe (error $ "role: unmatched type" <> show (typeRep @c))
             Just Refl  -> Just MutableAccount
             Nothing -> Nothing
 
+roleIndex :: forall (c :: Role). Typeable c => DerivationIndex
+roleIndex = fromMaybe (error $ "roleIndex: unmatched type" <> show (typeRep @c))
+       (tryUtxoExternal <|> tryUtxoInternal <|> tryMutableAccount)
+  where
+    tryUtxoExternal =
+        case testEquality (typeRep @c) (typeRep @'UtxoExternal) of
+            Just Refl  -> Just (DerivationIndex 0)
+            Nothing -> Nothing
+    tryUtxoInternal =
+        case testEquality (typeRep @c) (typeRep @'UtxoInternal) of
+            Just Refl  -> Just (DerivationIndex 1)
+            Nothing -> Nothing
+    tryMutableAccount =
+        case testEquality (typeRep @c) (typeRep @'MutableAccount) of
+            Just Refl  -> Just (DerivationIndex 2)
+            Nothing -> Nothing
```

Comment:
Seems like this is duplicating logic with the `Enum` instance for `Role` already. I'd suggest instead to define a functon to reify a type-level role into a `Role` value, and then piggy-back on the Enum instance to get the role index to avoid duplicating bits of logic.